source/unified-test-format/unified-test-format.rst

```diff
@@ -195,6 +195,22 @@ error if an attempt is made to store an entity with a name that already exists
 in the map and MUST raise an error if an entity is not found for a name or is
 found but has an unexpected type.
 
+Test runners MUST provide a mechanism to retrieve entities from the entity
+map prior to the clearing of the entity map, as discussed in
+`Executing a Test`_. There MUST be a way to:
+
+- Retrieve an entity by its name (for example, to support retrieving the
+  iteration count stored by the ``storeIterationsAsEntity`` option.
+  
+- Retrieve all entities of a particular type as a mapping of each entity name
+  to the respective documents (for example, to support retrieving
+  the event documents stored by the ``storeEventsAsEntities`` option).
```

Comment:
In the case of `storeEventsAsEntities`, doesn't that create one or more list/array entities to which events will be appended? If so, wouldn't the previous bullet be sufficient?

Given the following example:

```
storeEventsAsEntities:
  events: [PoolCreatedEvent, ConnectionCreatedEvent]
```

The workload executor would just need to access the "events" entity to obtain the list of serialized event documents.

Reply:
This is true, the issue is that workload executor doesn't know that the entity is called `events`.

Reply:
Understood, but I don't see why we need an API to fetch "all entities of type X" from the entity map to address this.

If the unified test exists within the `driverWorkload` key of the workload file (e.g. `retryReads-testFailover.yml`) you can standardize on the name there. I don't believe a YAML alias/reference can be used as a key, so you can't do:

```yml
storeEventsAsEntities:
  *eventListEntityName: [ PoolCreatedEvent, ConnectionCreatedEvent]]
```

However, you could have a separate option (possible sibling of `driverWorkload`) that the executor can read to infer the entity name to be fetched. For example:

```yml
cmapEventsEntity: events

...

driverWorkload:
  ...
    storeEventsAsEntities:
      events: [ PoolCreatedEvent, ConnectionCreatedEvent]
```

Alternatively:

 * the executor could decide to peek into the `driverWorkload` document itself and try to read `storeEventsAsEntities` for the client(s)
 * you can specify exactly what entity names `driverWorkload` configurations should use for capturing each type of CMAP event, and then executors can reliably assume where to look

That seems more foolproof than attempting to fetch, for example, all BSON arrays in the entity map. IIRC, drivers aren't obligated to define an "event list" entity type. Separately, if workload authors decide they'd like to split up pool and connection events into separate lists, I think the executor itself would have an easier time deciding to directly query for both lists explicitly instead of fetching all lists and then having to differentiate them from one another by inspecting the elements within.

Reply:
I think this is a fair point. I think for now I'll just mandate specific entity names in the DAT scenarios and in the future the flexible names could be implemented on the DAT side. I removed the requirement to retrieve the mapping and added an option to constrain retrieved entities with their expected type.